pkg/storage/mysql/mysqlstore.go

```diff
@@ -0,0 +1,268 @@
+	key   string
+	value []byte
+}
+
+type queryParams struct {
+	StartKey string `json:"startKey"`
+	EndKey   string `json:"endKey"`
+}
+
+const (
+	blankDBPathErrMsg         = "DB URL for new mySQL DB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+	tablePrefix               = "t_"
+	sqlDBNotFound             = "no rows"
+	createDBQuery             = "CREATE DATABASE IF NOT EXISTS "
+	useDBQuery                = "USE "
+)
+
+// NewProvider instantiates Provider
+func NewProvider(dbPath, keyLength string) (*Provider, error) {
+	if dbPath == "" {
+		return nil, errors.New(blankDBPathErrMsg)
+	}
+
+	// Example DB Path root:my-secret-pw@tcp(127.0.0.1:3306)/
+	db, err := sql.Open("mysql", dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open connection: %w", err)
+	}
+
+	p := &Provider{
+		dbURL: dbPath,
+		db:    db,
+		dbs:   map[string]*sqlDBStore{}}
+
+	return p, nil
+}
+
+// OpenStore opens and returns new db for given name space.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	// check name is not empty
+	if name == "" {
+		return nil, errors.New("store name is required")
+	}
+	// creating the database
+	_, err := p.db.Exec(createDBQuery + name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create db %s: %w", name, err)
+	}
+
+	// Opening new db connection
+	newDBConn, err := sql.Open("mysql", p.dbURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection %s: %w", p.dbURL, err)
+	}
+
+	// Use is used to select the created database without this DDL operations are not permitted
+	_, err = newDBConn.Exec(useDBQuery + name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to use db %s: %w", name, err)
+	}
+
+	tableName := tablePrefix + name
+	// key has max varchar size it can accommodate as per mysql 8.0 spec
+	createTableStmt := "CREATE Table IF NOT EXISTS " + tableName +
+		"(`key` varchar(255) NOT NULL ,`value` BLOB, PRIMARY KEY (`key`));"
+
+	// creating key-value table inside the database
+	_, err = newDBConn.Exec(createTableStmt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create table %s: %w", tableName, err)
+	}
+
+	store := &sqlDBStore{
+		db:        newDBConn,
+		tableName: tableName}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, store := range p.dbs {
+		err := store.db.Close()
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	if err := p.db.Close(); err != nil {
+		return err
+	}
+
+	p.dbs = make(map[string]*sqlDBStore)
+
+	return nil
+}
+
+// CloseStore closes a previously opened store
+func (p *Provider) CloseStore(name string) error {
+	p.Lock()
+	defer p.Unlock()
+
+	store, exists := p.dbs[name]
+	if !exists {
+		return nil
+	}
+
+	delete(p.dbs, name)
+
+	return store.db.Close()
+}
+
+// Put stores the key and the record
```

Comment:
Record generally refers to the entire row of data - I think you mean "value"

(keep in mind that our stores here are really just (key, value) pairs)